### PR TITLE
Fixed #35173 -- Reallowed filtering with __isnull on foreign keys not listed in list_filters.

### PR DESCRIPTION
**Pull request by Hisham-Pak**
Ticket: https://code.djangoproject.com/ticket/35173

### CODE REVIEW

**Comment by nessita** [comment]
Hello @Hisham-Pak, thank you for this pull request.

I don't think this is the right approach: as far as I see this branch is just restoring the whole `if` guard that was removed in https://code.djangoproject.com/changeset/f80669d2f5a5f1db9e9b73ca893fefba34f955e7, which sounds like a disproportionate solution for the reported problem.

Ideally, the fix would only enlarge/modify the current `if` guard to cover for the missing case. Or, the branch should come up with a solution that would replace the current `if` branch, but without resorting to having two apparently independent `if` guards that are really duplicating some of the checks.

Does this make sense? cc/ @sarahboyce

**Comment by Hisham-Pak** [comment]
Hi @nessita, thanks for the review. Currently I was not able to find a way to combine these checks in one `if` because of a break in between:
```
if not getattr(field, "path_infos", None):
    # This is not a relational field, so further parts
    # must be transforms.
    break
```
Will update this if I found a way around.

**Comment by nessita** [comment]
> Hi @nessita, thanks for the review. Currently I was not able to find a way to combine these checks in one `if` because of a break in between:
> 
> ```
> if not getattr(field, "path_infos", None):
>     # This is not a relational field, so further parts
>     # must be transforms.
>     break
> ```
> 
> Will update this if I found a way around. And yes I think `not prev_field` could be removed from second `if`.

Thank you for the response. If the `break` is causing the issue, you could consider reverting the current `if` to what it had before the regression commit but in a manner that the fixed issue in PR #17581, which has already a regression test, keeps passing.

**Comment by Hisham-Pak** [comment]
@nessita, I implemented the changes you suggested. Let me know if it looks good.

**Review by harshitbudhiraja — APPROVED** [review]

**Review by sarahboyce — APPROVED** [review]
Thank you for working on this @Hisham-Pak

File: django/contrib/admin/options.py [hunk 1]
```diff
@@ -475,24 +475,25 @@ def lookup_allowed(self, lookup, value, request=None):
...
                 break
+            if not prev_field or (
+                prev_field.is_relation
+                and field not in model._meta.parents.values()
+                and field is not model._meta.auto_field
+                and (
+                    model._meta.auto_field is None
+                    or part not in getattr(prev_field, "to_fields", [])
+                )
+                and (field.is_relation or not getattr(field, "primary_key", None))
```
Comment (nessita) [inline django/contrib/admin/options.py:486]: I have a question about this: the `not getattr(field, "primary_key", None)`, what is it covering exactly? is it asking for whether there is primary_key field? or if the primaty_key field is None? what happens if the value of `primary_key` is falsy?

I change the code to be:
```suggestion
                and (field.is_relation or not field.primary_key)
```
and no test fails, so I think that we can either simplify the code or add tests ensuring the have the more robust guard here.

Comment (Hisham-Pak) [reply django/contrib/admin/options.py]: `not getattr(field, "primary_key", None)` was checking for `not(hasattr(field, "primary_key") and field.primary_key)`. But I think this was not necessary as all fields inherit from `Field` class which already has `primary_key` attribute. The only exception I can think is of reverse relations and for that `field.is_relation` is already set to True.

Comment (nessita) [reply django/contrib/admin/options.py]: Thanks @Hisham-Pak, would you have time to extend the test case or add a new one with such case so we evaluate if we need to further tweak the `if` guard?

Comment (Hisham-Pak) [reply django/contrib/admin/options.py]: I don't think we need any further tweaking. It already is looking good.

**Pull request merged**